File: src/main/java/me/breidenbach/rabbitex/connection/RabbitConnection.java
package me.breidenbach.rabbitex.connection;

import com.rabbitmq.client.AMQP;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import me.breidenbach.rabbitex.Consumer;
import me.breidenbach.rabbitex.MessageHandler;
import me.breidenbach.rabbitex.Options;
import me.breidenbach.rabbitex.RabbitEx;

import java.io.IOException;
import java.util.Calendar;
import java.util.Map;

/**
 * User: Kevin E. Breidenbach
 * Date: 11/1/13
 * Time: 3:59 PM
 * © 2013 Kevin E. Breidenbach
 */
public class RabbitConnection implements RabbitEx {
    private static final boolean MANDATORY = true;
    private static final String EXCHANGE_TYPE = "topic";

    private final RabbitConnectionCache cache;
    private final ConnectionFactory factory;

    private final Connection connection; //the underlying RabbitConnection

    private boolean closed = false;

    RabbitConnection(final RabbitConnectionCache cache, final ConnectionFactory factory)
            throws RabbitConnectionException {
        this.cache = cache;
        this.factory = factory;
        this.connection = prepareRabbitConnection();
    }

    @Override
    public void close() throws IOException {
        connection.close();
        cache.remove(factory.getHost(), factory.getPort(), factory.getVirtualHost(), factory.getUsername());
        closed = true;
    }

    @Override
    public void publish(final String exchange, final String subject,
                        final String message, final Map<Options, String> options) throws RabbitConnectionException {
        MessageWrapper wrapper = createWrapper(message, MessageWrapper.MessageType.MESSAGE, options);
        publishMessage(exchange, subject, wrapper);
    }

    @Override
    public Consumer consumer(final String exchange, final String subject,
                             final String queue, final MessageHandler handler) throws RabbitConnectionException {
        return null;
    }

    public boolean isClosed() {
        return closed;
    }

    void publishError(final String exchange, final String subject, final String message)
            throws RabbitConnectionException {
        MessageWrapper wrapper = createWrapper(message, MessageWrapper.MessageType.ERROR, null);
        publishMessage(exchange, subject, wrapper);
    }

    private void publishMessage(final String exchange, final String subject, final MessageWrapper wrapper)
            throws RabbitConnectionException {
        AMQP.BasicProperties.Builder builder =  new AMQP.BasicProperties.Builder();
        String json = wrapper.toJson();
        builder.deliveryMode(2);
        builder.timestamp(Calendar.getInstance().getTime());
        try {
            Channel channel = connection.createChannel();
            channel.exchangeDeclare(exchange, EXCHANGE_TYPE, true);
            channel.basicPublish(exchange, subject, MANDATORY, builder.build(), json.getBytes());
        } catch (IOException e) {
            throw new RabbitConnectionException("Unable to publish message", e);
        }
    }

    private Connection prepareRabbitConnection() throws RabbitConnectionException {
        try {
            return factory.newConnection();
        } catch (IOException e) {
            throw new RabbitConnectionException("Unable to create connection: " + e.getMessage(), e);
        }
    }

    private MessageWrapper createWrapper(final String message, final MessageWrapper.MessageType type,
                                         final Map<Options,String> options) {
        MessageWrapper wrapper = new MessageWrapper(message, type);
        if (options != null) {
            for (Options key: options.keySet()) {
                switch (key) {
                    case ERROR_EXCHANGE:
                        wrapper.setErrorExchange(options.get(key));
                        break;
                    case ERROR_SUBJECT:
                        wrapper.setErrorSubject(options.get(key));
                        break;
                }
            }
        }
        return wrapper;
    }
}

File: src/main/scala/me/breidenbach/rabbitex/connection/RabbitConnection.scala
package me.breidenbach.rabbitex.connection

import java.io.IOException
import java.util
import java.util.Calendar

import com.rabbitmq.client.{AMQP, Connection, ConnectionFactory}
import me.breidenbach.rabbitex.{Consumer, MessageHandler, Options, RabbitEx}
import me.breidenbach.rabbitex.Options._
import scala.collection.mutable

/**
 * Date: 10/12/14
 * Time: 7:14 PM
 * Copyright 2014 Kevin E. Breidenbach
 * @author Kevin E. Breidenbach
 */
private[connection] case class RabbitConnection(host: String, port: Int, virtualHost: String, username: String,
                                                password: String, factory: ConnectionFactory = null) extends RabbitEx {

  val key: String = host + port + virtualHost + username
  val connection = prepareConnection()
  var closed = false

  private val EXCHANGE_TYPE = "topic"

  override def publish(exchange: String, subject: String, message: String, options: util.Map[Options, String] = null): Unit = {
    val errorExchange = if (options != null && options.containsKey(ERROR_EXCHANGE)) Some(options.get(ERROR_EXCHANGE)) else None
    val errorSubject = if (options != null && options.containsKey(ERROR_SUBJECT)) Some(options.get(ERROR_SUBJECT)) else None
    val wrapper: MessageWrapper = Message(message = message, errorExchange = errorExchange, errorSubject = errorSubject)
    publish(exchange, subject, wrapper)
  }

  def publishError(exchange: String, subject: String, errorAction: HandlerResponse.HandlerResponse, message: String): Unit = {
    val wrapper: MessageWrapper = ErrorMessage(message, errorAction)
    publish(exchange, subject, wrapper)
  }

  def publish(exchange: String, subject: String, wrapper: MessageWrapper): Unit = {
    val builder =  new AMQP.BasicProperties.Builder()
    val json = wrapper.toJson
    builder.deliveryMode(2)
    builder.timestamp(Calendar.getInstance().getTime)
    try {
      val channel = connection.createChannel()
      channel.exchangeDeclare(exchange, EXCHANGE_TYPE, true)
      channel.basicPublish(exchange, subject, builder.build(), json.getBytes)
    } catch {
      case e: IOException => throw new RabbitConnectionException("Unable to publish message", e);
    }
  }

  override def consumer(exchange: String, subject: String, queue: String, handler: MessageHandler): Consumer = {
    RabbitConsumer(this, exchange, subject, queue, handler)
  }

  override def close(): Unit = {
    RabbitConnection.remove(this)
    closed = true
  }

  private def prepareConnection(): Connection = factory match {
    case null => null
    case _ =>
      try {
        factory.newConnection()
      }
      catch {
        case e: IOException =>
          throw new RabbitConnectionException("Unable to create connection: " + e.getMessage, e)
      }
  }
}

object RabbitFactory {
  private implicit val factory = new ConnectionFactory

  def newConnection(host: String, port: Int, virtualHost: String = "", username: String = "", password: String = ""): RabbitEx = {
    RabbitConnection.newConnection(host, port, virtualHost, username, password)
  }
}

object RabbitConnection {

  private val connections = new mutable.WeakHashMap[String, RabbitConnection]

  // The implicit factory enables testing
  def newConnection(host: String, port: Int, virtualHost: String = "", username: String = "", password: String = "")
                   (implicit factory: ConnectionFactory): RabbitEx = {
    val conn = new RabbitConnection(host, port, virtualHost, username, password)
    connections.contains(conn.key) match {
      case true =>
        val connection = connections(conn.key)
        if (connection.closed) cacheNewConnection(conn) else connection
      case _ =>
        cacheNewConnection(conn)
    }
  }

  private[connection] def remove(rabbitConnection: RabbitConnection): Unit = {
    connections.remove(rabbitConnection.key)
  }

  private def cacheNewConnection(rabbitConnection: RabbitConnection)(implicit factory: ConnectionFactory): RabbitEx = {
    factory.setHost(rabbitConnection.host)
    factory.setPort(rabbitConnection.port)
    if (rabbitConnection.username != null && !rabbitConnection.username.isEmpty) factory.setUsername(rabbitConnection.username)
    if (rabbitConnection.password != null && !rabbitConnection.password.isEmpty) factory.setPassword(rabbitConnection.password)

    val rabbitExConnection = rabbitConnection.copy(factory = factory)
    connections(rabbitExConnection.key) = rabbitExConnection
    rabbitExConnection
  }
}


